Lagrange: user feedback features

darominski opened the pull request:
**Description of work.**

This PR implements issues and requests reported by the scientific user, namely:
- [x] Bug: Water correction does not take into account the difference between energy exchange or initial energy situations leading to interpolation going to the region with no data has been addressed consistently with the sample data.
- [x] More workspaces are now be available for debugging, notably the water correction, both pre- and post-interpolation
- [x] User now has the ability to choose normalisation from: None, and Monitor. Time will be added after the switch to NXS.
- [x] Metadata of time and temperature is now loaded from the ASCII and added to the sample logs.

The support for loading NXS files will be added as soon as the format is decided. 

**To test:**
1. Code review.
2. All tests should be passing
3. Run the example added to the documentation (make sure you have LAGRANGE unit test data in your search directories):
```
# full correction of a single monochromator scan, with multiple files
result = LagrangeILLReduction(SampleRuns='012869:012871',
                                  ContainerRuns='012882:012884',
                                  CorrectionFile='correction-water-cu220-2020.txt',
                                  UseIncidentEnergy=False,
                                  ConvertToWaveNumber=False)
 run = result.getRun()
 times = run.getLogData('time').value
 eis = result.readX(0)
 temperatures = run.getLogData('temperature').value
 plt.plot(eis, temperatures)
 plt.xlabel("Ei (meV)")
 plt.ylabel("temperature (K)")
 plt.show()
```
A proper figure should appear with temperature as a function of initial energy displayed correctly. There should be a rapid change at the lowest energies and plateau towards high-Ei.

Fixes #34837. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
[LAGRANGE] Added water correction to intermediate workspaces

including interpolated water correction, so that both can be accessed while debugging.

darominski committed:
[LAGRANGE] Added offset option to water correction

in case the energy transfer is used instead of initial energy for the sample.
This ensures consistency of ranges between the water correction workspace and the sample.

darominski committed:
[LAGRANGE] Added option to choose normalisation approach

one from: None, or Monitor.

darominski committed:
[LAGRANGE] Corrected settin interpolated water ws name

and fixed incorrect type hints in the same method.

darominski committed:
[LAGRANGE] Moved monitor_erro fetching for error propagation

darominski committed:
[LAGRANGE] Added loading time and temperature metadata from ASCII

darominski committed:
[LAGRANGE] Release note for water, normalisation, metadata update

gui-co commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/LagrangeILLReduction.py:
```diff
@@ -232,6 +289,16 @@ def get_water_correction(self, correction_file: str) -> np.ndarray:
+            self.intermediate_workspaces.append(water_corr_ws)
         return correction
 
     def merge_adjacent_points(self, data: np.ndarray) -> np.ndarray:
```
What happens to metadata (temperature and time) when merging adjacent points?

darominski replied:
That's a very good point. I changed this behaviour to mask the input `ndarray` so that the temperature and time is also masked for points that are not taken into account.

gui-co reviewed (commented):
The code is working as expected. I am just wondering how the merge of metadata should be handled in `merge_adjacent_points()`.

darominski committed:
[LAGRANGE] Proper masking of merged points

Now the points that are deemed to be too close in Ei are masked out, which also takes care of the loaded metadata.

gui-co commented:
Thanks for the changes and for the additional test that covers this edge case.

I also realized that the display of array of values is not great in the samples logs when the array is long. This is not really related to this, I opened a specific PR for this: #34870

sf1919 commented:
We had to temporarily disable the linux and system tests pipeline. Now it's up again I've set a job for this PR running at https://builds.mantidproject.org/job/pull_requests-conda-linux/2282/

Please note: the link to this job will not be picked up by the checks list. It will still show as pending. Providing the job linked to above passes please review as normal.

gui-co commented:
Thanks for this.
There will be new commits to this PR and a new build will be triggered.

darominski committed:
[LAGRANGE] Modified naming scheme for intermediate workspaces

and fixed an issue with not updating empty container (EC) name to be subtracted from the sample after
water correction has been applied to EC. Previously, the correction has been applied to EC, but the uncorrected workspace has been subtracted.

gui-co reviewed (dismissed):
This is now ready to be merged.

darominski committed:
[LAGRANGE] Added loading and setting of run_title

darominski commented:
Apologies @gui-co for dismissing the review all this time, this is the third time I am getting the 'final' comment from the user. Since they are leaving for Christmas, this is the last one this year and this concludes this part of the work. Thank you for your patience.

gui-co commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/LagrangeILLReduction.py:
```diff
@@ -44,6 +44,9 @@ class LagrangeILLReduction(DataProcessorAlgorithm):
     normalise_by = None
 
+    # workspace title
+    title = None
```
The name is not consistent with the property you are using later: `self._title`
